powerscale/provider/user_mapping_rules_resource.go

```diff
@@ -0,0 +1,513 @@
+									MarkdownDescription: "Specifies the user name.",
+									Computed:            true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// Configure configures the resource.
+func (r *UserMappingRulesResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	pscaleClient, ok := req.ProviderData.(*client.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = pscaleClient
+}
+
+// Create allocates the resource.
+func (r *UserMappingRulesResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	tflog.Info(ctx, "Creating User Mapping Rules resource state")
+	// Read Terraform plan into the model
+	var plan models.UserMappingRulesResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var state models.UserMappingRulesResourceModel
+	if !plan.Rules.IsUnknown() || !plan.Parameters.IsUnknown() {
+		if diags := helper.UpdateUserMappingRules(ctx, r.client, &state, &plan); diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+	}
+
+	rulesResponse, err := helper.GetUserMappingRulesByZone(ctx, r.client, plan.Zone.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("error getting user mapping rules", err.Error())
+		return
+	}
+
+	if diags := helper.UpdateUserMappingRulesState(ctx, &plan, rulesResponse); diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	if diags := helper.UpdateLookupMappingUsersState(ctx, r.client, &plan); diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	tflog.Info(ctx, "Done with Create User Mapping Rules resource state")
+}
+
+// Read reads the resource state.
+func (r *UserMappingRulesResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	tflog.Info(ctx, "Reading User Mapping Rules resource state")
+
+	var state models.UserMappingRulesResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	rulesResponse, err := helper.GetUserMappingRulesByZone(ctx, r.client, state.Zone.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("error getting user mapping rules", err.Error())
+		return
+	}
+
+	if diags := helper.UpdateUserMappingRulesState(ctx, &state, rulesResponse); diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	if diags := helper.UpdateLookupMappingUsersState(ctx, r.client, &state); diags.HasError() {
```

Comment:
@taohe1012 Hi Tao , could you help me understand what is lookup mapping user? What's its relations with user mapping rules

Reply:
@RayLiu7 , according to onefs ui, the user mapping rules also provide ways to check test user mapping result.
In tf, you can input list users for user mapping test.
![image](https://github.com/dell/terraform-provider-powerscale/assets/88763781/d5af1faa-a7af-4a1d-96c1-3bbc3d51de3f)

Reply:
Thanks for clarifying!

Reply:
I am wondering if `test_mapping_users` should be a part of resource `powerscale_user_mapping_rules`...

Reply:
@baoy1  when adding rules for users, the onefs won't check the users are valid or invalid, so it's one way to check mapping rules. And as above screenshot,  it's also one part of the user mapping rules page.

Reply:
Terraform resource block should focus on the infrastructure or resource management. It appears to me a validation method makes no sense to be included in a resource block. How about querying mapping rule via datasource?

Reply:
@baoy1  same for datasource, datasource is to query for rules detail, if we add validation function, it's also weird. So thinking put it in resource should also be possible.